include/nbl/builtin/hlsl/bxdf/bxdf_traits.hlsl

```diff
@@ -0,0 +1,107 @@
+struct traits;
+
+template<typename L, typename I, typename A, typename S>
+struct traits<bxdf::reflection::SLambertianBxDF<L, I, A, S> >
+{
+    NBL_CONSTEXPR_STATIC_INLINE BxDFType type = BT_BRDF;
+    NBL_CONSTEXPR_STATIC_INLINE bool clampNdotV = false;
+    NBL_CONSTEXPR_STATIC_INLINE bool clampNdotL = true;
+};
+
+template<typename L, typename I, typename A, typename S>
+struct traits<bxdf::reflection::SOrenNayarBxDF<L, I, A, S> >
+{
+    NBL_CONSTEXPR_STATIC_INLINE BxDFType type = BT_BRDF;
+    NBL_CONSTEXPR_STATIC_INLINE bool clampNdotV = true;
+    NBL_CONSTEXPR_STATIC_INLINE bool clampNdotL = true;
+};
+
+// no blinn phong
+
+template<typename L, typename I, typename A, typename IC, typename AC, typename S>
+struct traits<bxdf::reflection::SBeckmannBxDF<L, I, A, IC, AC, S> >
+{
+    NBL_CONSTEXPR_STATIC_INLINE BxDFType type = BT_BRDF;
+    NBL_CONSTEXPR_STATIC_INLINE bool clampNdotV = true;
+    NBL_CONSTEXPR_STATIC_INLINE bool clampNdotL = true;
+};
+
+template<typename L, typename I, typename A, typename IC, typename AC, typename S>
+struct traits<bxdf::reflection::SGGXBxDF<L, I, A, IC, AC, S> >
+{
+    NBL_CONSTEXPR_STATIC_INLINE BxDFType type = BT_BRDF;
+    NBL_CONSTEXPR_STATIC_INLINE bool clampNdotV = true;
+    NBL_CONSTEXPR_STATIC_INLINE bool clampNdotL = true;
+};
+
+
+template<typename L, typename I, typename A, typename S>
+struct traits<bxdf::transmission::SLambertianBxDF<L, I, A, S> >
+{
+    NBL_CONSTEXPR_STATIC_INLINE BxDFType type = BT_BSDF;
+    NBL_CONSTEXPR_STATIC_INLINE bool clampNdotV = false;
+    NBL_CONSTEXPR_STATIC_INLINE bool clampNdotL = true;
+};
+
+template<typename L, typename I, typename A, typename IC, typename AC, typename S>
+struct traits<bxdf::transmission::SSmoothDielectricBxDF<L, I, A, IC, AC, S, false> >
+{
+    NBL_CONSTEXPR_STATIC_INLINE BxDFType type = BT_BSDF;
+    NBL_CONSTEXPR_STATIC_INLINE bool clampNdotV = true;
+    NBL_CONSTEXPR_STATIC_INLINE bool clampNdotL = true;
+};
+
+template<typename L, typename I, typename A, typename IC, typename AC, typename S>
+struct traits<bxdf::transmission::SSmoothDielectricBxDF<L, I, A, IC, AC, S, true> >
+{
+    NBL_CONSTEXPR_STATIC_INLINE BxDFType type = BT_BSDF;
+    NBL_CONSTEXPR_STATIC_INLINE bool clampNdotV = true;
+    NBL_CONSTEXPR_STATIC_INLINE bool clampNdotL = true;
+};
+
+template<typename L, typename I, typename A, typename IC, typename AC, typename S>
+struct traits<bxdf::transmission::SBeckmannDielectricBxDF<L, I, A, IC, AC, S> >
+{
+    NBL_CONSTEXPR_STATIC_INLINE BxDFType type = BT_BSDF;
+    NBL_CONSTEXPR_STATIC_INLINE bool clampNdotV = true;
+    NBL_CONSTEXPR_STATIC_INLINE bool clampNdotL = true;
+};
+
+template<typename L, typename I, typename A, typename IC, typename AC, typename S>
+struct traits<bxdf::transmission::SGGXDielectricBxDF<L, I, A, IC, AC, S> >
+{
+    NBL_CONSTEXPR_STATIC_INLINE BxDFType type = BT_BSDF;
+    NBL_CONSTEXPR_STATIC_INLINE bool clampNdotV = true;
+    NBL_CONSTEXPR_STATIC_INLINE bool clampNdotL = true;
+};
```

Comment:
the partial specializations of the traits should appear RIGHT after the definition of each struct (its easy to avoid template instantiation before specialization that way, and you don't have to go looking)